src/common/host_device_vector.cu

```diff
@@ -17,6 +17,16 @@ struct HostDeviceVectorImpl {
       data_h_.resize(size);
     }
   }
+  HostDeviceVectorImpl(std::initializer_list<T> init, int device)
```

Comment:
I like how you are supporting `std::initializer_list<T>` so that it's possible to use braced array constants to initialize `HostDeviceVector`. Here is a suggestion: why don't we also suggest initialization from `std::vector<T>` as well? This way, users can initialize `HostDeviceVector` with constants as well as vectors.

To minimize code duplication, we can have `HostDeviceVectorImpl` constructor to accept `std::vector<T>` and then expose two interfaces from `HostDeviceVector` that constructs `HostDeviceVectorImpl` objects.
```cpp
template <typename T>
HostDeviceVector<T>::HostDeviceVector(const std::vector<T>& init, int device)
  : impl_(new HostDeviceVectorImpl<T>(init)) {}

template <typename T>
HostDeviceVector<T>::HostDeviceVector(std::initializer_list<T> init, int device)
  : impl_(new HostDeviceVectorImpl<T>(init)) {}

...

template <typename T>
struct HostDeviceVectorImpl {
  HostDeviceVectorImpl(size_t size, T val, int device) { ... }
  HostDeviceVectorImpl(const std::vector<T>& init, int device) { ... }
};
```

Reply:
Done.